File: src/hooksIngest.ts
import { existsSync, readFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import * as db from "./db.js";

export const HOOK_EVENTS_PATH = join(homedir(), ".claude", "skills-usage", "hook-events.jsonl");

interface HookEvent {
  received_at?: string;
  hook_event_name?: string;
  tool_use_id?: string;
  duration_ms?: number;
}

export function ingestHookEvents(
  path: string = HOOK_EVENTS_PATH,
  dbPath: string = db.DB_PATH,
): { events: number; updated: number; skipped: number } {
  if (!existsSync(path)) return { events: 0, updated: 0, skipped: 0 };

  const preAt: Record<string, string> = {};
  const postAt: Record<string, string> = {};
  const durations: Record<string, number> = {};

  const raw = readFileSync(path, "utf-8");
  let events = 0;
  for (const line of raw.split("\n")) {
    if (!line.trim()) continue;
    let e: HookEvent;
    try { e = JSON.parse(line); } catch { continue; }
    events += 1;
    const tid = e.tool_use_id;
    if (!tid) continue;
    if (e.hook_event_name === "PreToolUse") {
      const ra = e.received_at ?? "";
      if (!(tid in preAt) || ra < preAt[tid]) preAt[tid] = ra;
    } else if (e.hook_event_name === "PostToolUse") {
      const ra = e.received_at ?? "";
      if (!(tid in postAt) || ra > postAt[tid]) postAt[tid] = ra;
      if (typeof e.duration_ms === "number") durations[tid] = e.duration_ms;
    }
  }

  let updated = 0;
  let skipped = 0;
  const allIds = new Set([...Object.keys(preAt), ...Object.keys(postAt)]);
  for (const tid of allIds) {
    const t0 = preAt[tid] ?? null;
    const t1 = postAt[tid] ?? null;
    let durSec: number | null = null;
    if (tid in durations) {
      durSec = durations[tid] / 1000;
    } else if (t0 && t1) {
      const p0 = Date.parse(t0);
      const p1 = Date.parse(t1);
      if (!Number.isNaN(p0) && !Number.isNaN(p1)) durSec = (p1 - p0) / 1000;
    }
    if (db.updateHookTiming(tid, t0, t1, durSec, dbPath)) updated += 1;
    else skipped += 1;
  }

  return { events, updated, skipped };
}

File: archive-py/skills_usage/hooks_ingest.py
"""Ingest hook-events.jsonl produced by hooks/skill_event.py and join into DB."""
from __future__ import annotations

import json
from datetime import datetime
from pathlib import Path

from . import db as dbmod

HOOK_EVENTS_PATH = Path.home() / ".claude" / "skills-usage" / "hook-events.jsonl"


def _parse_iso(s: str | None):
    if not s:
        return None
    try:
        return datetime.fromisoformat(s.replace("Z", "+00:00"))
    except Exception:
        return None


def ingest_hook_events(path: Path = HOOK_EVENTS_PATH, db_path: Path = dbmod.DB_PATH) -> tuple[int, int, int]:
    """Read hook-events.jsonl and update skill_calls with hook_started_at/ended_at/duration.

    Returns (events_read, calls_updated, calls_skipped_no_match).
    """
    if not path.exists():
        return 0, 0, 0

    # Group events by tool_use_id
    pre_at: dict[str, str] = {}
    post_at: dict[str, str] = {}
    durations: dict[str, int] = {}

    events = 0
    with open(path, encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                e = json.loads(line)
            except Exception:
                continue
            events += 1
            tid = e.get("tool_use_id")
            if not tid:
                continue
            ev = e.get("hook_event_name")
            if ev == "PreToolUse":
                # Earliest Pre wins
                if tid not in pre_at or e.get("received_at", "") < pre_at[tid]:
                    pre_at[tid] = e.get("received_at", "")
            elif ev == "PostToolUse":
                # Latest Post wins
                if tid not in post_at or e.get("received_at", "") > post_at[tid]:
                    post_at[tid] = e.get("received_at", "")
                if (d := e.get("duration_ms")) is not None:
                    durations[tid] = int(d)

    updated = skipped = 0
    conn = dbmod._connect(db_path)
    try:
        cur = conn.cursor()
        all_ids = set(pre_at) | set(post_at)
        for tid in all_ids:
            t0 = pre_at.get(tid)
            t1 = post_at.get(tid)
            dur_sec = None
            if tid in durations:
                dur_sec = durations[tid] / 1000.0
            elif t0 and t1:
                p0, p1 = _parse_iso(t0), _parse_iso(t1)
                if p0 and p1:
                    dur_sec = (p1 - p0).total_seconds()

            res = cur.execute(
                "UPDATE skill_calls SET hook_started_at=?, hook_ended_at=?, hook_duration_sec=?, updated_at=CURRENT_TIMESTAMP "
                "WHERE tool_use_id=?",
                (t0, t1, dur_sec, tid),
            )
            if res.rowcount > 0:
                updated += 1
            else:
                skipped += 1
        conn.commit()
    finally:
        conn.close()

    return events, updated, skipped


if __name__ == "__main__":
    e, u, s = ingest_hook_events()
    print(f"hook events: {e}, calls updated: {u}, skipped (no transcript yet): {s}")
